torch/lib/THD/base/data_channels/DataChannelGloo.cpp

```diff
@@ -66,32 +66,42 @@ void DataChannelGloo::RequestGloo::wait() {
 DataChannelGloo::DataChannelGloo()
   : _rank(load_rank_env())
 {
-  _store = std::unique_ptr<::gloo::rendezvous::Store>(new thd::Store());
-
   if (_rank == 0) {
     _num_processes = load_world_size_env();
-    auto num_proc_str = std::to_string(_num_processes);
-    _store->set("world_size",
-            std::vector<char>(num_proc_str.begin(), num_proc_str.end()));
-  } else {
-    auto world_size = _store->get("world_size");
-    _num_processes = std::atoll(
-            std::string(world_size.begin(), world_size.end()).c_str());
   }
 
-  ::gloo::transport::tcp::attr attr; // default options listen on all interfaces
+  ::gloo::transport::tcp::attr attr("localhost"); // default options listen on host
   _device = ::gloo::transport::tcp::CreateDevice(attr);
 
-  GlooCache::get().setRank(_rank);
-  GlooCache::get().setDevice(_device);
+  // Master runs the store_thread
+  if (_rank == 0) {
+    std::tie(_port, std::ignore) = load_master_env();
+    _addr = "localhost";
+  } else {
+    std::tie(_addr, _port) = load_worker_env();
+  }
 }
 
 
-DataChannelGloo::~DataChannelGloo() {
-}
+DataChannelGloo::~DataChannelGloo() {}
 
 
 bool DataChannelGloo::init() {
+  _store = std::unique_ptr<::gloo::rendezvous::Store>(
+    new thd::Store(_rank, _addr, _port, _num_processes)
+  );
+  _cache = std::unique_ptr<GlooCache>(new GlooCache(_rank, _device, _store));
+
+  if (_rank == 0) {
+    auto num_proc_str = std::to_string(_num_processes);
+    _store->set("world_size",
+            std::vector<char>(num_proc_str.begin(), num_proc_str.end()));
+  } else {
+    auto world_size = _store->get("world_size");
+    _num_processes = std::atoll(
+            std::string(world_size.begin(), world_size.end()).c_str());
+  }
+
   std::vector<rank_type> ranks;
   ranks.reserve(_num_processes);
```

Comment:
Unfortunately, it cannot. Because `std::vector<rank_type> ranks(_num_processes)` not only allocates a vector's storage but also fills it with default values whereas `reserve` only allocates storage.

Summing up:
`reserve` changes capacity but not the size of a vector.
`std::vector<rank_type> ranks(_num_processes)` changes capacity AND size of a vector.